### DISCUSSION TITLE
Utilizing appended Stabilizers

### DISCUSSION THREAD

**Discussion by Fe-r-oz** [QuantumSavory/QuantumClifford.jl#303] [Q&A] [answered]
A very trivial question:

```
julia> s = parity_checks(Steane7())
+ ___XXXX
+ _XX__XX
+ X_X_X_X
+ ___ZZZZ
+ _ZZ__ZZ
+ Z_Z_Z_Z

julia> hcat(s, s)
1×2 Matrix{QuantumClifford.Stabilizer{QuantumClifford.Tableau{Vector{UInt8}, Matrix{UInt64}}}}:
 Stabilizer 6×7  Stabilizer 6×7

tried something like this as well:
julia> parity_checks(hcat(s, s))
ERROR: MethodError: no method matching parity_checks(::Matrix{QuantumClifford.Stabilizer{QuantumClifford.Tableau{Vector{UInt8}, Matrix{UInt64}}}})

Closest candidates are:
  parity_checks(::QuantumClifford.ECC.Toric)
   @ QuantumClifford ~/GitHub/QuantumClifford/QuantumClifford.jl/src/ecc/codes/toric.jl:46
  parity_checks(::QuantumClifford.ECC.ReedMuller)
   @ QuantumClifford ~/GitHub/QuantumClifford/QuantumClifford.jl/src/ecc/codes/classical/reedmuller.jl:30
  parity_checks(::QuantumClifford.ECC.GottesmanDistance4)
   @ QuantumClifford ~/GitHub/QuantumClifford/QuantumClifford.jl/src/ecc/codes/gottesmandistance4.jl:17
  ...

Stacktrace:
 [1] top-level scope
   @ REPL[6]:1

```

I am unable to access or see this appended matrix in a Stabilizer form. Can we append (horizontally) first 3 rows, all columns then  0 then first 3 rows and all columns and use such appended matrix, meaning in pseudocode` append( s[1:3, 1:7] 0 s[1:3, 1:7])`  and then utilize it?

Why is this important? Some quantum codes may contain either X-Stabilizers or Z-Stabilizers of smaller quantum codes with some appending. I was thinking maybe we can test it combining it with canonicalize! since canonical representations should be equivalent. This is not necessarily a stringent requirement as better tests can be made that tests mathematical properties which cover entire family of qcode rather than geometric-like property which only cover specific member of family of qcode.

**Comment by Krastanov** [comment]
Do you mean one of these:

```
julia> s1 = random_stabilizer(4)
+ X_XZ
+ X_ZX
- ZX_Y
+ _YXZ

julia> s2 = random_stabilizer(4)
+ _XYY
+ ZYXY
+ _YYX
- ZXYY

julia> s1⊗s2
+ X_XZ____
+ X_ZX____
- ZX_Y____
+ _YXZ____
+ _____XYY
+ ____ZYXY
+ _____YYX
- ____ZXYY

julia> vcat(s1,s2)
+ X_XZ
+ X_ZX
- ZX_Y
+ _YXZ
+ _XYY
+ ZYXY
+ _YYX
- ZXYY

```

**Reply by Fe-r-oz** [reply]
Thanks! These are helpful but I am looking for a different scenario in which X-stabilizers of a shorter code is appended horizontally twice in a specific way (for example) to form the larger code.

Suppose we have this following code
[[15, 1, 3]] quantum Reed-Muller code

```
julia> parity_checks(QuantumReedMuller(4))
+ X_X_X_X_X_X_X_X
+ _XX__XX__XX__XX
+ ___XXXX____XXXX
+ _______XXXXXXXX
+ Z___Z___Z___Z__
+ _Z___Z___Z___Z_
+ __Z___Z___Z___Z
+ ___ZZZZ____ZZZZ
+ ____Z_Z_____Z_Z
+ _____ZZ______ZZ
+ _______ZZZZZZZZ
+ ________Z_Z_Z_Z
+ _________ZZ__ZZ
+ ___________ZZZZ

```
The first three rows of [[15, 1, 7]] QRM code has X Stabilizers of Steane7(), 0, X Stabilizers of Steane7() appended horizontally. I was thinking to test it somehow. 

```
julia> canonicalize!(parity_checks(Steane7()))
+ X_X_X_X
+ _XX__XX
+ ___XXXX
+ Z_Z_Z_Z
+ _ZZ__ZZ
+ ___ZZZZ
```

**Reply by Fe-r-oz** [reply]
Same is the case with the Z-Stabilizers. Two copies of Z-stabilizers with 0 appended in between. I read that this 15 qubit QRM  code has 2 copies of `Steane7() `code :) which can be seen visually.  But I would like to have test for it. 

Extract from the 15 qubit QRM code
```
+ X_X_X_X_X_X_X_X
+ _XX__XX__XX__XX
+ ___XXXX____XXXX
+ ___ZZZZ____ZZZZ
+ ____Z_Z_____Z_Z
+ _____ZZ______ZZ
```

This way we can see the 2 copies Steane7() with 0/I in between clearly. A separate question: Does it make sense to rearrange the stabilizer like this, maybe for testing purpose ?
```
julia> parity_checks(QuantumReedMuller(4))
+ _______XXXXXXXX
+ X_X_X_X_X_X_X_X  #append(Steane(7) 0 Steane7())  - here
+ _XX__XX__XX__XX  
+ ___XXXX____XXXX 
+ ___ZZZZ____ZZZZ 
+ ____Z_Z_____Z_Z  
+ _____ZZ______ZZ   #append(Steane(7) 0 Steane7())  - to here
+ Z___Z___Z___Z__
+ _Z___Z___Z___Z_
+ __Z___Z___Z___Z
+ _______ZZZZZZZZ
+ ________Z_Z_Z_Z
+ _________ZZ__ZZ
+ ___________ZZZZ
```

**Reply by Krastanov** [reply, marked as answer]
I do not think an `hcat` method is implemented, but it makes sense to create one. If you want to, start a PR for it and I can help finish it.

**Reply by Fe-r-oz** [reply]
Sure! I found the `vcat` method as a reference:

https://github.com/QuantumSavory/QuantumClifford.jl/blob/dddaedbac7ac76ae6ce4fb1fc171acdc142a9b53/src/QuantumClifford.jl#L916-L923
